### ISSUE TITLE
investigate chroma changed filtering behavior

### ISSUE THREAD

**Issue by pmeier** [Quansight/ragna#523]
In be943a75fd8027e77fb14d75ebe2a8a830d4cba7 as part of #461 I changed the behavior of LanceDB to behave more like Chroma when filtering results. See https://github.com/Quansight/ragna/pull/461#issuecomment-2275737866 for details. However, I think upstream Chroma might have aligned with LanceDB in the meantime. If it weren't for the upper pin

https://github.com/Quansight/ragna/blob/71acf83d4114514c56e298f7acc431551933e522/ragna/source_storages/_vector_database.py#L49

we see some test failures. I couldn't find a changelog or issue indicating the change in behavior.

We should align Chroma and LanceDB on the latest version again. That might mean that we need to revert the patch mentioned above and fix the tests.

**Comment by smokestacklightnin** [comment]
As an experiment, I tried updating the chromadb version. Most, if not all, of the test failures come from the same cause. The test failure mentions 

```
NotImplementedError: In Chroma v0.6.0, list_collections only returns collection names. Use Client.get_collection(default) to access query. See https://docs.trychroma.com/deployment/migration for more information.
```

The [migration guide (v0.6.0)](https://docs.trychroma.com/production/administration/migration#migration-log) discusses how to correctly use `list_collections` and get collection names in the current version. I tried this on one of the test failures and it started passing. I assume this solution will work for all failures of this type.

All that said, we can and probably should do some cleanup on the filtering code

**Comment by pmeier** [comment]
Indeed `chromadb>=0.6` brings a whole lot of new issues and we should follow their upgrade guide. However, these are _not_ the issues that are relevant here. If you `pip install --upgrade 'chromadb>=0.5.12,<0.6'` you get

```shell
$ pytest -k chroma
[...]
======================================== FAILURES =========================================
______________________________ test_smoke[Chroma-or-nested] _______________________________
tests/source_storages/test_source_storages.py:137: in test_smoke
    assert actual_idcs == expected_idcs
E   assert [0, 1, 2] == [0, 1]
E     
E     Left contains one more item: 2
E     Use -v to get more diff
__________________________________ test_smoke[Chroma-ne] __________________________________
tests/source_storages/test_source_storages.py:137: in test_smoke
    assert actual_idcs == expected_idcs
E   assert [2, 3, 4, 5, 6] == [2, 5, 6]
E     
E     At index 1 diff: 3 != 5
E     Left contains 2 more items, first extra item: 5
E     Use -v to get more diff
________________________________ test_smoke[Chroma-not_in] ________________________________
tests/source_storages/test_source_storages.py:137: in test_smoke
    assert actual_idcs == expected_idcs
E   assert [0, 1, 2, 3, 4] == [0, 1, 2]
E     
E     Left contains 2 more items, first extra item: 3
E     Use -v to get more diff
```

**Comment by smokestacklightnin** [comment]
[These three test failures](https://github.com/Quansight/ragna/issues/523#issuecomment-2567459980) seem to come from what is discussed in [v0.5.12 from the migration guide](https://docs.trychroma.com/production/administration/migration#migration-log). I looked at all three cases and the changes there are the cause of the discrepancies. 

For convenience, here is the description of the change:

> The operators `$ne` (not equal) and `$nin` (not in) in `where` clauses have been updated:
>
> - Previously: They only matched records that had the specified key.
> - Now: They also match records that don't have the specified key at all.
>
> In other words, `$ne` and `$nin` now match the complement set of records (the exact opposite) that `$eq` (equals) and `$in` (in) would match, respectively.
>
> The `$not_contains` operator in the `where_document` clause has also been updated:
>
> - Previously: It only matched records that had a document field.
> - Now: It also matches records that don't have a document field at all.
>
> In other words, `$not_contains` now matches the exact opposite set of records that `$contains` would match.

**Comment by pmeier** [comment]
Thanks for the pointer. This is what I suspected in the first place. Thus, we just need to rollback the patch on `LanceDB` and should be good to go.

**Comment by pmeier** [comment]
@anush008 Coming from #524: how does Qdrant handle missing keys when filtering data?

**Comment by Anush008** [comment]
Hey @pmeier.
If you mean, what happens when a filter is applied on a key that's not present in data, the filter simply fails to match.

**Comment by pmeier** [comment]
Yeah that is what I suspected since the tests in #524 are passing. However this diverges from what LanceDB is doing and also what Chroma is doing as of a few releases (see https://github.com/Quansight/ragna/issues/523#issuecomment-2568802732). This issue is about aligning the behavior of the builtin source storages with the common ground now being matching also happens on the negative operators ("not equal", "not in") if the key is not present at all. Meaning, we'll have to update #524 as well. Test failures will show up there when a PR is merged that closes this issue.

**Comment by smokestacklightnin** [comment]
It appears that Chroma is the outlier among itself, Lance, and Qdrant with respect to the [behavior for missing keys with the `ne` and `nin` metadata filter operators](https://github.com/Quansight/ragna/issues/523#issuecomment-2568802732). Unfortunately, I am not finding (and don't want sink too much time into it) a way to make Chroma behave like the other two. The issue is that the [metadata filtering operators](https://docs.trychroma.com/docs/querying-collections/metadata-filtering) don't allow for filtering based on the presence or absence of keys for the operators we are looking at.

**Comment by smokestacklightnin** [comment]
I'll also add that I think we should keep the [typecasting null columns in Lance](https://github.com/Quansight/ragna/blob/main/ragna/source_storages/_lancedb.py#L167) because otherwise we run into issues with typing in tests.

**Comment by pmeier** [comment]
Kinda unfortunate that Chroma was aligned with the other source storages and decided to diverge. Even more so that there is no option for us to maintain the old behavior. But there is little we can do now. I agree that we should align with LanceDB and Qdrant from #524. Thus, let's upgrade Chroma to >=0.6, fix the tests, and add a fat disclaimer to the docs that Chroma filtering behavior diverges for non-existing keys.
